Tribes/src/app/_services/building.service.ts

```diff
@@ -27,14 +28,47 @@ export class BuildingService {
   }
 
   addNewBuilding(type: string): void {
+    this.updateRessourceByConstruction.emit(COST_NEW_BUILDING);
     this.http.post<Building>(ROOT_URL + '/kingdom/buildings', {type})
     .pipe(catchError(this.errorHandlingService.handleError))
     .subscribe(response => this.handleBuildingProcess(response),
               error => console.error(error));
   }
 
-  handleBuildingProcess(building: Building) {
+  handleBuildingProcess(building: Building): void {
     this.beginConstruction.emit();
-    setTimeout(() => {this.finishConstruction.emit(); }, building.finishedAt - building.startedAt);
+    localStorage.setItem('finishedAt' + building.id, building.finishedAt.toString());
+    setTimeout(() => { this.finishConstruction.emit(); }, building.finishedAt - building.startedAt);
+  }
+
+  upgradeBuilding(idToUpgrade: number, level: number): void {
+    this.updateRessourceByConstruction.emit(COST_BASE_UPGRADE * (level - 1));
+    this.http.put<Building>(ROOT_URL + '/kingdom/buildings/' + idToUpgrade, { level })
+    .pipe(catchError(this.errorHandlingService.handleError))
+    .subscribe(response => this.handleBuildingProcess(response),
+              error => console.error(error));
+  }
+
+  checkIfBuildingIsProgressing(building: Building): boolean {
+    const currentTime = new Date().getTime();
+    return currentTime <= building.finishedAt;
+  }
+
+  setTimeoutsAgain() {
+    if (localStorage.length > 1) {
+      for (let i = 0; i < localStorage.length; i++) {
+        let key: string = localStorage.key(i);
+        if (key !== 'TOKEN') {
```

Comment:
You most definitely do not have to iterate over the items :)

You can just do something like this:
```
// To save:
const MY_KEY = 'someprefix_is_illdomos'
localStorage.setItem(MY_KEY, JSON.stringify(['timeout1', 'timeout2', ...]))
// To load:
let timeoutList = []
if (localStorage.getItem(MY_KEY)){
timeOutList = JSON.parse(localStorage.getItem('app_timeouts'))}
```